File: domain/src/test/java/no/nav/melosys/domain/dokument/organisasjon/OrganisasjonsDetaljerTest.java
package no.nav.melosys.domain.dokument.organisasjon;

import java.util.Arrays;

import no.nav.melosys.domain.dokument.adresse.StrukturertAdresse;
import no.nav.melosys.domain.dokument.adresse.UstrukturertAdresse;
import no.nav.melosys.domain.dokument.felles.Periode;
import no.nav.melosys.domain.dokument.organisasjon.adresse.SemistrukturertAdresse;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class OrganisasjonsDetaljerTest {

    private SemistrukturertAdresse adresse;

    private String linje1 = "LINJE1  ";
    private String linje2 = "LINJE2";
    private String linje3 = "LINJE3";
    private String postnr = "postnummer";
    private String poststed = "poststed";
    private String poststedUtland = "poststedUtland";

    @BeforeEach
    public void setUp() {

        Periode periode = mock(Periode.class);
        when(periode.erGyldig()).thenReturn(true);

        adresse = new SemistrukturertAdresse();
        adresse.setAdresselinje1(linje1);
        adresse.setAdresselinje2(linje2);
        adresse.setAdresselinje3(linje3);
        adresse.setPostnr(postnr);
        adresse.setPoststed(poststed);
        adresse.setPoststedUtland(poststedUtland);
        String kommunenr = "kommunenr";
        adresse.setKommunenr(kommunenr);
        adresse.setGyldighetsperiode(periode);
    }

    @Test
    public void testKonverterForretningsadresseTilUstrukturertAdresse() {
        String landkode = "NO";
        adresse.setLandkode(landkode);

        OrganisasjonsDetaljer orgDetaljer = new OrganisasjonsDetaljer();
        orgDetaljer.forretningsadresse = Arrays.asList(adresse);

        UstrukturertAdresse resultatAdresse = orgDetaljer.hentUstrukturertForretningsadresse();
        assertThat(resultatAdresse.getAdresselinje(1)).isEqualTo(linje1);
        assertThat(resultatAdresse.getAdresselinje(2)).isEqualTo(linje2);
        assertThat(resultatAdresse.getAdresselinje(3)).isEqualTo(linje3);
        assertThat(resultatAdresse.getAdresselinje(4)).isEqualTo(postnr + " " + poststed);
        assertThat(resultatAdresse.landkode).isEqualTo(landkode);
    }

    @Test
    public void testKonverterUtenlandskForretningsadresseTilUstrukturertAdresse() {
        String landkode = "DK";
        adresse.setLandkode(landkode);

        OrganisasjonsDetaljer orgDetaljer = new OrganisasjonsDetaljer();
        orgDetaljer.forretningsadresse = Arrays.asList(adresse);

        UstrukturertAdresse resultatAdresse = orgDetaljer.hentUstrukturertForretningsadresse();
        assertThat(resultatAdresse.getAdresselinje(1)).isEqualTo(linje1);
        assertThat(resultatAdresse.getAdresselinje(2)).isEqualTo(linje2);
        assertThat(resultatAdresse.getAdresselinje(3)).isEqualTo(linje3);
        assertThat(resultatAdresse.getAdresselinje(4)).isEqualTo(poststedUtland);
        assertThat(resultatAdresse.landkode).isEqualTo(landkode);
    }

    @Test
    public void testKonverterForretningsadresseTilStrukturertAdresse() {
        String landkode = "NO";
        adresse.setLandkode(landkode);

        OrganisasjonsDetaljer orgDetaljer = new OrganisasjonsDetaljer();
        orgDetaljer.forretningsadresse = Arrays.asList(adresse);

        StrukturertAdresse resultatAdresse = orgDetaljer.hentStrukturertForretningsadresse();
        assertThat(resultatAdresse.gatenavn).isEqualTo(linje1.trim() + " " + linje2 + " " + linje3);
        assertThat(resultatAdresse.landkode).isEqualTo(landkode);

        assertThat(resultatAdresse.postnummer).isEqualTo(postnr);
        // Ikke alltid poststed for norske registeradresser. Slåes opp med kodeverkservice ved behov
        assertThat(resultatAdresse.poststed).isEqualTo(poststed);
    }

    @Test
    public void testKonverterUtenlandskForretningsadresseTilStrukturertAdresse() {
        String landkode = "DK";
        adresse.setLandkode(landkode);

        OrganisasjonsDetaljer orgDetaljer = new OrganisasjonsDetaljer();
        orgDetaljer.forretningsadresse = Arrays.asList(adresse);

        StrukturertAdresse resultatAdresse = orgDetaljer.hentStrukturertForretningsadresse();
        assertThat(resultatAdresse.gatenavn).isEqualTo(linje1.trim() + " " + linje2 + " " + linje3);
        assertThat(resultatAdresse.landkode).isEqualTo(landkode);
        assertThat(resultatAdresse.postnummer).isEqualTo(postnr);
        assertThat(resultatAdresse.poststed).isEqualTo(poststedUtland);
    }

    @Test
    public void testNullAdresse() {
        OrganisasjonsDetaljer orgDetaljer = new OrganisasjonsDetaljer();
        StrukturertAdresse resultatAdresse = orgDetaljer.hentStrukturertForretningsadresse();

        assertThat(resultatAdresse).isNull();
    }
}

File: domain/src/test/java/no/nav/melosys/domain/dokument/organisasjon/OrganisasjonsDetaljerTest.kt
//package no.nav.melosys.domain.dokument.organisasjon
//
//import no.nav.melosys.domain.dokument.felles.Periode
//import no.nav.melosys.domain.dokument.organisasjon.adresse.GeografiskAdresse
//import no.nav.melosys.domain.dokument.organisasjon.adresse.SemistrukturertAdresse
//import org.assertj.core.api.Assertions
//import org.junit.jupiter.api.BeforeEach
//import org.junit.jupiter.api.Test
//import org.mockito.Mockito
//import java.util.*
//
//class OrganisasjonsDetaljerTest {
//    private var adresse: SemistrukturertAdresse? = null
//    private val linje1: String = "LINJE1  "
//    private val linje2: String = "LINJE2"
//    private val linje3: String = "LINJE3"
//    private val postnr: String = "postnummer"
//    private val poststed: String = "poststed"
//    private val poststedUtland: String = "poststedUtland"
//    @BeforeEach
//    fun setUp() {
//        val periode = Mockito.mock(
//            Periode::class.java
//        )
//        Mockito.`when`(periode.erGyldig()).thenReturn(true)
//        adresse = SemistrukturertAdresse()
//        adresse.setAdresselinje1(linje1)
//        adresse.setAdresselinje2(linje2)
//        adresse.setAdresselinje3(linje3)
//        adresse.setPostnr(postnr)
//        adresse.setPoststed(poststed)
//        adresse.setPoststedUtland(poststedUtland)
//        val kommunenr = "kommunenr"
//        adresse.setKommunenr(kommunenr)
//        adresse.setGyldighetsperiode(periode)
//    }
//
//    @Test
//    fun testKonverterForretningsadresseTilUstrukturertAdresse() {
//        val landkode = "NO"
//        adresse.setLandkode(landkode)
//        val orgDetaljer = OrganisasjonsDetaljer()
//        orgDetaljer.forretningsadresse = Arrays.asList<GeografiskAdresse?>(adresse)
//        val resultatAdresse = orgDetaljer.hentUstrukturertForretningsadresse()
//        Assertions.assertThat(resultatAdresse.getAdresselinje(1)).isEqualTo(linje1)
//        Assertions.assertThat(resultatAdresse.getAdresselinje(2)).isEqualTo(linje2)
//        Assertions.assertThat(resultatAdresse.getAdresselinje(3)).isEqualTo(linje3)
//        Assertions.assertThat(resultatAdresse.getAdresselinje(4)).isEqualTo("$postnr $poststed")
//        Assertions.assertThat(resultatAdresse.landkode).isEqualTo(landkode)
//    }
//
//    @Test
//    fun testKonverterUtenlandskForretningsadresseTilUstrukturertAdresse() {
//        val landkode = "DK"
//        adresse.setLandkode(landkode)
//        val orgDetaljer = OrganisasjonsDetaljer()
//        orgDetaljer.forretningsadresse = Arrays.asList<GeografiskAdresse?>(adresse)
//        val resultatAdresse = orgDetaljer.hentUstrukturertForretningsadresse()
//        Assertions.assertThat(resultatAdresse.getAdresselinje(1)).isEqualTo(linje1)
//        Assertions.assertThat(resultatAdresse.getAdresselinje(2)).isEqualTo(linje2)
//        Assertions.assertThat(resultatAdresse.getAdresselinje(3)).isEqualTo(linje3)
//        Assertions.assertThat(resultatAdresse.getAdresselinje(4)).isEqualTo(poststedUtland)
//        Assertions.assertThat(resultatAdresse.landkode).isEqualTo(landkode)
//    }
//
//    @Test
//    fun testKonverterForretningsadresseTilStrukturertAdresse() {
//        val landkode = "NO"
//        adresse.setLandkode(landkode)
//        val orgDetaljer = OrganisasjonsDetaljer()
//        orgDetaljer.forretningsadresse = Arrays.asList<GeografiskAdresse?>(adresse)
//        val resultatAdresse = orgDetaljer.hentStrukturertForretningsadresse()
//        Assertions.assertThat(resultatAdresse.gatenavn)
//            .isEqualTo(linje1.trim { it <= ' ' } + " " + linje2 + " " + linje3)
//        Assertions.assertThat(resultatAdresse.landkode).isEqualTo(landkode)
//        Assertions.assertThat(resultatAdresse.postnummer).isEqualTo(postnr)
//        // Ikke alltid poststed for norske registeradresser. Slåes opp med kodeverkservice ved behov
//        Assertions.assertThat(resultatAdresse.poststed).isEqualTo(poststed)
//    }
//
//    @Test
//    fun testKonverterUtenlandskForretningsadresseTilStrukturertAdresse() {
//        val landkode = "DK"
//        adresse.setLandkode(landkode)
//        val orgDetaljer = OrganisasjonsDetaljer()
//        orgDetaljer.forretningsadresse = Arrays.asList<GeografiskAdresse?>(adresse)
//        val resultatAdresse = orgDetaljer.hentStrukturertForretningsadresse()
//        Assertions.assertThat(resultatAdresse.gatenavn)
//            .isEqualTo(linje1.trim { it <= ' ' } + " " + linje2 + " " + linje3)
//        Assertions.assertThat(resultatAdresse.landkode).isEqualTo(landkode)
//        Assertions.assertThat(resultatAdresse.postnummer).isEqualTo(postnr)
//        Assertions.assertThat(resultatAdresse.poststed).isEqualTo(poststedUtland)
//    }
//
//    @Test
//    fun testNullAdresse() {
//        val orgDetaljer = OrganisasjonsDetaljer()
//        val resultatAdresse = orgDetaljer.hentStrukturertForretningsadresse()
//        Assertions.assertThat(resultatAdresse).isNull()
//    }
//}
